### PR TITLE
Sockets: Shubha & Grace

### PR DESCRIPTION
**Pull request by gracemshea**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | Method Signature which includes: def, method name, parameters. This is followed by a block of code.
What are the advantages of using git when collaboratively working on one code base? | It allows for much easier version control and prevents us from overriding each other's work when editing from different machines.
What kind of relationship did you and your pair have with the unit tests? | For each method, we started with the tests failing, and accomplished the revisions necessary to make it pass. And then we proceeded to the next method!
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We did. We used the reduce method in our word_score section (Line 58). It allowed us to succinctly add together the scores of different characters as we iterated through the words.
What was one method you and your pair used to debug code? | We used binding.pry to access the values of variables at break points.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We were both comfortable stopping the other when we needed to slow down or clarify a question. We discussed our communication styles and while we work very well together - we did discuss how we would approach working with "guess" communication styles.

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+# Returns an array of 10 letters chosen at random from available letters
+def draw_letters
+  letter_freq = {
+    A: 9, N: 6, B: 2, O: 8, C: 2, P: 2, D: 4, Q: 1, E: 12, R: 6, F: 2, S: 4,
+    G: 3, T: 6, H: 2, U: 4, I: 9, V: 2, J: 1, W: 2, K: 1, X: 1, L: 4, Y: 2, M: 2, Z: 1,
+  }
+
+  avail_letters = array_gen(letter_freq)
+  used_letters = avail_letters.sample(10)
```
Comment (dHelmgren) [inline lib/adagrams.rb:22]: This is lovely! I can see everything I need to see about how this works, and if I get confused about `array_gen` it's close at hand!

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+# Validates input word by determining if it can be formed with letters in hand
+def uses_available_letters?(input, letters_in_hand)
+  if !(input.is_a?(String))
+    raise ArgumentError, "Ummmmmm the value for the first argument needs to be a string, ok? Given value: #{input}"
+  elsif !(letters_in_hand.is_a?(Array))
+    raise ArgumentError, "The second argument should be an array. That doesn't look right...Given value: #{letters_in_hand}"
+  end
+  if input.length > letters_in_hand.length
+    return false
+  else
```
Comment (dHelmgren) [inline lib/adagrams.rb:36]: Small semantic note: If you don't `return` above or `raise`, then you won't hit the code in this `else` anyway. In cases like this, it's nice to leave the meat of the method outside of an `else` block.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    # reassigning letters_in_hand to new variable in order to avoid destruction of original array
+    possible_letters = letters_in_hand.clone
+    input.upcase.split(//).each do |char|
+      if possible_letters.include?(char)
+        possible_letters.delete(char)
```
Comment (dHelmgren) [inline lib/adagrams.rb:41]: You will never return `true` for a word with a repeated letter. For example,
    ```ruby
    hand = ["E", "F", "R", "A", "E", "L", "V", "A", "D", "K"]
    uses_available_letters?("ADA", hand)
    # => false
    ```
    The reason why is that Ruby's `Array#delete` method will remove all matching elements from an array, not just the first match.

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      else
+        return false
```
Comment (dHelmgren) [inline lib/adagrams.rb:43]: Nice work returning when you have the answer!

**Comment by dHelmgren** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Yes
Both teammates contributed to the codebase | Yes
Small commits with meaningful commit messages | Yes
**Code Requirements** |
`draw_letters` method | Yes
Uses appropriate data structure to store the letter distribution | Yes
All tests for `draw_letters` pass | Yes
`uses_available_letters?` method | See comment!
All tests for `uses_available_letters?` pass | Yes
`score_word` method | Yes
Uses appropriate data structure to store the letter scores | Yes, also, Nice Memo!
All tests for `score_word` pass | Yes
`highest_score_from` method | Yes
Appropriately handles edge cases for tie-breaking logic | Yes
All tests for `highest_score_from` pass | Yes
**Overall** | Hey y'all! Great work on this project. Your code fulfills all of the requirements and does it in a readable and reasonable way. The code is clean, and it passes all the tests!

File: lib/adagrams.rb (hunk 3 above)
Comment (shubha-rajan) [reply lib/adagrams.rb:41]: Good to know! 
If we were to write something like:
`i = index (char)`
`possible_letters.delete_at(i) `
would it preserve the other repeats of the letter?
And is there a more elegant way of doing it?
